File: ui/article-renderer/src/main/java/org/cru/godtools/article/adapter/CategoriesAdapter.java
package org.cru.godtools.article.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.ccci.gto.android.common.recyclerview.adapter.SimpleDataBindingAdapter;
import org.cru.godtools.article.databinding.ListItemCategoryBinding;
import org.cru.godtools.xml.model.Category;

import java.util.List;

public class CategoriesAdapter extends SimpleDataBindingAdapter<ListItemCategoryBinding> {
    public interface Callbacks {
        void onCategorySelected(@Nullable final Category category);
    }

    @Nullable
    private List<Category> mCategories;

    @Nullable
    private Callbacks mCallbacks;

    @Override
    public int getItemCount() {
        return mCategories != null ? mCategories.size() : 0;
    }

    public void setCategories(@Nullable final List<Category> categories) {
        mCategories = categories;
        notifyDataSetChanged();
    }

    public void setCallbacks(@Nullable final Callbacks callbacks) {
        mCallbacks = callbacks;
        notifyItemRangeChanged(0, getItemCount());
    }

    // region Lifecycle Events

    @NonNull
    @Override
    protected ListItemCategoryBinding onCreateViewDataBinding(@NonNull final ViewGroup parent, final int viewType) {
        return ListItemCategoryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
    }

    @Override
    protected void onBindViewDataBinding(@NonNull final ListItemCategoryBinding binding, final int position) {
        assert mCategories != null : "mCategories has to be defined to trigger onBindViewDataBinding";
        binding.setCallbacks(mCallbacks);
        binding.setCategory(mCategories.get(position));
    }

    @Override
    protected void onViewDataBindingRecycled(@NonNull final ListItemCategoryBinding binding) {
        super.onViewDataBindingRecycled(binding);
        binding.setCallbacks(null);
        binding.setCategory(null);
    }

    // endregion Lifecycle Events
}

File: ui/article-renderer/src/main/java/org/cru/godtools/article/adapter/CategoriesAdapter.kt
package org.cru.godtools.article.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.ObservableField
import androidx.recyclerview.widget.RecyclerView.NO_ID
import org.ccci.gto.android.common.recyclerview.adapter.SimpleDataBindingAdapter
import org.ccci.gto.android.common.support.v4.util.IdUtils
import org.cru.godtools.article.databinding.ListItemCategoryBinding
import org.cru.godtools.xml.model.Category

class CategoriesAdapter : SimpleDataBindingAdapter<ListItemCategoryBinding>() {
    interface Callbacks {
        fun onCategorySelected(category: Category?)
    }

    init {
        setHasStableIds(true)
    }

    private val callbacks = ObservableField<Callbacks>()
    var categories: List<Category>? = null
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    override fun getItemCount(): Int {
        return categories?.size ?: 0
    }

    override fun getItemId(position: Int): Long {
        return categories?.get(position)?.id?.let { IdUtils.convertId(it) } ?: NO_ID
    }

    fun setCallbacks(callbacks: Callbacks?) = this.callbacks.set(callbacks)

    // region Lifecycle Events

    override fun onCreateViewDataBinding(parent: ViewGroup, viewType: Int): ListItemCategoryBinding {
        return ListItemCategoryBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            .also { it.callbacks = callbacks }
    }

    override fun onBindViewDataBinding(binding: ListItemCategoryBinding, position: Int) {
        binding.category = categories?.get(position)
    }

    override fun onViewDataBindingRecycled(binding: ListItemCategoryBinding) {
        super.onViewDataBindingRecycled(binding)
        binding.category = null
    }

    // endregion Lifecycle Events
}
